File: src/test/java/org/dicio/skill/FallbackSkillTest.java
package org.dicio.skill;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertSame;

import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;

import org.dicio.skill.chain.InputRecognizer;
import org.junit.Test;

import java.util.List;

public class FallbackSkillTest {

    @Test
    public void testConstructor() {
        new FallbackSkill() {
            @Override public void setInput(final String input, final List<String> inputWords,
                    final List<String> normalizedWordKeys) {}
            @Override public void processInput() {}
            @Override public void generateOutput() {}
            @Override public void cleanup() {}
        };
    }

    @Test
    public void testScoreAndSpecificity() {
        final Skill skill = new FallbackSkill() {
            @Override public void setInput(final String input, final List<String> inputWords,
                    final List<String> normalizedWordKeys) {}
            @Override public void processInput() {}
            @Override public void generateOutput() {}
            @Override public void cleanup() {}
        };

        assertEquals(0.0f, skill.score(), 0.0f);
        assertSame(InputRecognizer.Specificity.low, skill.specificity());
    }
}

File: src/test/java/org/dicio/skill/FallbackSkillTest.kt
package org.dicio.skill

import io.kotest.core.spec.style.StringSpec
import io.kotest.matchers.shouldBe
import io.kotest.matchers.types.shouldBeSameInstanceAs
import org.dicio.skill.chain.InputRecognizer.Specificity

class FallbackSkillTest : StringSpec({
    "constructor" {
        object : FallbackSkill(TestSkillInfo) {
            override fun setInput(
                input: String, inputWords: List<String>,
                normalizedWordKeys: List<String>
            ) {
            }

            override fun processInput() {}
            override fun generateOutput() = throw NotImplementedError()
            override fun cleanup() {}
        }
    }

    "score and specificity" {
        val skill: Skill = object : FallbackSkill(TestSkillInfo) {
            override fun setInput(
                input: String, inputWords: List<String>,
                normalizedWordKeys: List<String>
            ) {
            }

            override fun processInput() {}
            override fun generateOutput() = throw NotImplementedError()
            override fun cleanup() {}
        }

        skill.score() shouldBe 0.0f
        skill.specificity shouldBeSameInstanceAs Specificity.LOW
    }
})

object TestSkillInfo : SkillInfo("", 0, 0, 0, false) {
    override fun isAvailable(context: SkillContext) = TODO()
    override fun build(context: SkillContext) = TODO()
    override val preferenceFragment get() = TODO()
}
